krabs/krabs/provider.hpp

```diff
@@ -10,6 +10,7 @@
 
```

Comment:
This definitely feels like we should either split this into another class `threadsafe_provider` or just make this one thread-safe and force the new API in a major version update.
At the very least, a bunch of the code dupe could be cleaned up with a CRTP pattern.
If you want an example of how the CRTP code would look, let me know.

Reply:
Went down the major version update / API breaking change path. 
Waiting to hear from @swannman if this is his preference.

Reply:
Yes, I am okay with a breaking change + major version update.